woff2otf.py

```diff
@@ -19,92 +19,124 @@
+            {
+                "tag": struct.unpack(">I", infile.read(4))[0],
+                "offset": struct.unpack(">I", infile.read(4))[0],
+                "compLength": struct.unpack(">I", infile.read(4))[0],
+                "origLength": struct.unpack(">I", infile.read(4))[0],
+                "origChecksum": struct.unpack(">I", infile.read(4))[0],
+            }
+        )
+        offset += 4 * 4
+
+    for TableDirectoryEntry in TableDirectoryEntries:
+        outfile.write(struct.pack(">I", TableDirectoryEntry["tag"]))
+        outfile.write(struct.pack(">I", TableDirectoryEntry["origChecksum"]))
         outfile.write(struct.pack(">I", offset))
-        outfile.write(struct.pack(">I", TableDirectoryEntry['origLength']))
-        TableDirectoryEntry['outOffset'] = offset
-        offset += TableDirectoryEntry['origLength']
+        outfile.write(struct.pack(">I", TableDirectoryEntry["origLength"]))
+        TableDirectoryEntry["outOffset"] = offset
+        offset += TableDirectoryEntry["origLength"]
         if (offset % 4) != 0:
             offset += 4 - (offset % 4)
-            
+
     for TableDirectoryEntry in TableDirectoryEntries:
-        infile.seek(TableDirectoryEntry['offset'])
-        compressedData = infile.read(TableDirectoryEntry['compLength'])
-        if TableDirectoryEntry['compLength'] != TableDirectoryEntry['origLength']:
+        infile.seek(TableDirectoryEntry["offset"])
+        compressedData = infile.read(TableDirectoryEntry["compLength"])
+        if TableDirectoryEntry["compLength"] != TableDirectoryEntry["origLength"]:
             uncompressedData = zlib.decompress(compressedData)
         else:
             uncompressedData = compressedData
-        outfile.seek(TableDirectoryEntry['outOffset'])
+        outfile.seek(TableDirectoryEntry["outOffset"])
         outfile.write(uncompressedData)
-        offset = TableDirectoryEntry['outOffset'] + TableDirectoryEntry['origLength'];
+        offset = TableDirectoryEntry["outOffset"] + TableDirectoryEntry["origLength"]
         padding = 0
         if (offset % 4) != 0:
             padding = 4 - (offset % 4)
-        outfile.write(bytearray(padding));
+        outfile.write(bytearray(padding))
 
 
-def convert(infilename, outfilename):
-    with open(infilename , mode='rb') as infile:
-        with open(outfilename, mode='wb') as outfile:
+def convert(in_filename: Union[str, Path], out_filename: Union[str, Path]):
+    with open(in_filename, mode="rb") as infile:
+        with open(out_filename, mode="wb") as outfile:
             convert_streams(infile, outfile)
 
 
-def main(argv):
+def discover_and_convert(source_dir: Path, target_dir: Path):
+    for file in source_dir.glob("*.woff"):
+        convert(file, target_dir / file.name.replace(".woff", ".otf"))
+    for subdir in source_dir.glob("**/"):  # first subdir is the dir itself
+        if subdir == source_dir:
+            continue
+        discover_and_convert(subdir, target_dir / subdir.name)
```

Comment:
Why the recursion and not just use `source.dir.glob("**/*.woff")` in the first place?

Reply:
Your question makes a lot of sense.
The only reason I opted for recursion is because I could keep the same directory structure in the output directory with no effort.

If the source directory looks like:

```shell
input_dir
    subdir1
        subdir11
            subdir111
                font.woff
```

With your approach and with no effort, the output dir would look like:

```shell
output_dir
    font.otf
```

If you have a clean way of keeping the subdir structure, I'd happily apply it.